pkg/agent/worker.go

```diff
@@ -0,0 +1,363 @@
+package agent
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pagent "github.com/livekit/protocol/agent"
+	"github.com/livekit/protocol/livekit"
+	"github.com/livekit/protocol/logger"
+	"github.com/livekit/protocol/utils"
+	putil "github.com/livekit/protocol/utils"
+)
+
+type WorkerProtocolVersion int
+
+const CurrentProtocol = 1
+
+const (
+	registerTimeout  = 10 * time.Second
+	assignJobTimeout = 10 * time.Second
+	pingFrequency    = 10 * time.Second
+)
+
+var (
+	ErrWorkerClosed        = errors.New("worker closed")
+	ErrWorkerNotAvailable  = errors.New("worker not available")
+	ErrAvailabilityTimeout = errors.New("agent worker availability timeout")
+)
+
+type Worker struct {
+	id          string
+	jobType     livekit.JobType
+	version     string
+	name        string
+	namespace   string
+	load        float32
+	permissions *livekit.ParticipantPermission
+	apiKey      string
+	apiSecret   string
+	serverInfo  *livekit.ServerInfo
+	mu          sync.Mutex
+
+	protocolVersion WorkerProtocolVersion
+	registered      atomic.Bool
+	status          livekit.WorkerStatus
+	runningJobs     map[string]*Job
+
+	onWorkerRegistered func(w *Worker)
+
+	msgChan chan *livekit.ServerMessage
+	closed  chan struct{}
+
+	availibility map[string]chan *livekit.AvailabilityResponse
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	Logger logger.Logger
+}
+
+func NewWorker(
+	protocolVersion WorkerProtocolVersion,
+	apiKey string,
+	apiSecret string,
+	serverInfo *livekit.ServerInfo,
+	logger logger.Logger,
+) *Worker {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	w := &Worker{
+		id:              putil.NewGuid(utils.AgentWorkerPrefix),
+		protocolVersion: protocolVersion,
+		apiKey:          apiKey,
+		apiSecret:       apiSecret,
+		serverInfo:      serverInfo,
+		closed:          make(chan struct{}),
+		runningJobs:     make(map[string]*Job),
+		availibility:    make(map[string]chan *livekit.AvailabilityResponse),
+		msgChan:         make(chan *livekit.ServerMessage),
+		ctx:             ctx,
+		cancel:          cancel,
+		Logger:          logger,
+	}
+
+	go func() {
+		<-time.After(registerTimeout)
+		if !w.registered.Load() && !w.IsClosed() {
+			w.Logger.Warnw("worker did not register in time", nil, "id", w.id)
+			w.Close()
+		}
+	}()
+
+	return w
+}
+
+func (w *Worker) sendRequest(req *livekit.ServerMessage) {
+	w.msgChan <- req
+}
+
+func (w *Worker) ReadChan() <-chan *livekit.ServerMessage {
```

Comment:
Not sure how idiomatic it is to create this `ReadChan` method.
The motivation is that the WSSignalConnection is inside the service pkg, to avoid cyclic dependencies, the worker can't have this object in its field. So I'm communicating with the agentservice using this channel

Reply:
Not sure to understand, should I use routing.MessageSink?